### PR TITLE
feat(gazelle)!: Move the plugin to a separate workspace

### PR DESCRIPTION
**Pull request by aignas**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [X] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Previously the gazelle plugin was part of the same `WORKSPACE`. This means that we cannot have a separate `bzlmod` module and make the `rules_python` repo work with both, `bzlmod` and legacy dependency management systems.

Issue Number: #965

## What is the new behavior?

The `gazelle` plugin is now isolated with its own `WORKSPACE` file. Whilst making this I have also moved the plugin source code to a separate directory.

Summary:
- Move go.mod to gazelle.
- Move `gazelle` definition.
- ~Move the gazelle plugin to a separate folder, just like `bazel-skylib` does, which helps with naming of the externally visible targets.~ This is now #983.
- Fix file `distribution` for the gazelle module.
- Update the example test.
- Include rules_python_gazelle_plugin during integration tests
- Update ignored packages

## Does this PR introduce a breaking change?

- [X] Yes
- [ ] No

Steps that need to be taken:
- Add a new dependency to your `WORKSPACE` and change the import path for the plugin dependency setup.
- If you are using the gazelle binary from the plugin repo, use the new label `@rules_python_gazelle_plugin//python:gazelle_binary`.
- If you are using the gazelle language extension, use the new label `@rules_python_gazelle_plugin//python`.
- `gazelle_python.yaml` integrity gets changed for all users.

### CODE REVIEW

File: .bazelrc [hunk 1]
```diff
@@ -3,8 +3,8 @@
...
-query --deleted_packages=examples/build_file_generation,examples/build_file_generation/get_url,examples/bzlmod,examples/multi_python_versions,examples/multi_python_versions/libs/my_lib,examples/multi_python_versions/requirements,examples/multi_python_versions/tests,examples/pip_install,examples/pip_parse,examples/pip_parse_vendored,examples/pip_repository_annotations,examples/py_import,examples/relative_requirements,tests/compile_pip_requirements,tests/pip_repository_entry_points,tests/pip_deps
+build --deleted_packages=examples/build_file_generation,examples/build_file_generation/random_number_generator,examples/bzlmod,examples/multi_python_versions,examples/multi_python_versions/libs/my_lib,examples/multi_python_versions/requirements,examples/multi_python_versions/tests,examples/pip_install,examples/pip_parse,examples/pip_parse_vendored,examples/pip_repository_annotations,examples/py_import,examples/relative_requirements,tests/compile_pip_requirements,tests/pip_repository_entry_points,tests/pip_deps
```
Comment (chrislovecnm) [inline .bazelrc:6]: This is fixing a bug that I missed that is outside the scope of this PR.  Can we do this in a different PR?

File: examples/build_file_generation/WORKSPACE [hunk 2]
```diff
@@ -67,6 +67,11 @@ local_repository(
     path = "../..",
 )
 
+local_repository(
```
Comment (chrislovecnm) [inline examples/build_file_generation/WORKSPACE:70]: Can we add a comment on what local_repository does and how you don't want to use it when you create your own project?

File: gazelle/.gitignore [hunk 3]
```diff
@@ -0,0 +1,52 @@
+# Compiled Object files
```
Comment (chrislovecnm) [inline gazelle/.gitignore:1]: Do we need this here?  Don't we have an ignore file in the root directory?

File: gazelle/python/python_test.go [hunk 4]
```diff
@@ -38,9 +38,9 @@ import (
 )
 
 const (
-	extensionDir      = "gazelle/"
+	extensionDir      = "python/"
```
Comment (chrislovecnm) [inline gazelle/python/python_test.go:41]: Should we use ` os.PathSeparator` here?

**Review by chrislovecnm — CHANGES_REQUESTED** [review]
Look great.  I have some nitpicks that I noticed, but I like it!!  Working on the build_file_generation example I was wondering if bzmod and gazelle where working.

**Comment by alexeagle** [comment]
FYI @f0rmiga is on paternity leave, back next week I believe

**Comment by chrislovecnm** [comment]
@alexeagle who else can review it?

**Comment by alexeagle** [comment]
I think we should probably just wait for @f0rmiga to be back. This will need someone to invest a few hours.

As one way to make this easier to land, @aignas is it possible to split out a separate PR that does the directory rename so this doesn't show as a 433 files changed? I can't even get started on reviewing it at this size without setting aside a chunk of my day.

**Comment by aignas** [comment]
Hey @alexeagle, I also think that waiting for @f0rmiga is probably for the best. I can split it to several commits at the very least. I think the rename is necessary to make everything work as we are moving the `gazelle` rule usage to the `gazelle` folder and it makes things a little bit tricky.

What I can do to make it easier to review is to split it into multiple PRs introducing a breaking change:
* Do the folder rename.
* Add a new WORKSPACE.
* Whatever is left afterwards.

How does it sound to you?

**Comment by alexeagle** [comment]
Yeah I think the folder rename first will make it more manageable. Thanks!

**Comment by aignas** [comment]
@f0rmiga, I think once you merge #983, this PR should be much more approachable.

**Review by f0rmiga — APPROVED** [review]

**Comment by f0rmiga** [comment]
Thanks for addressing this, @aignas!

**Pull request merged**